### PR TITLE
Introduce modules/adsense datastore

### PR DESCRIPTION
**Pull request by felixarntz**
## Summary

Addresses issue #1247

## Relevant technical choices

Mostly related to PHP (all JS things are based on decisions made while implementing the Analytics datastore):
* There's a new `GET:tag-permission` datapoint inspired by how the one for Analytics works.
* The `GET:accounts` datapoint has an ugly side effect to store the first account as the account to use. We won't need this anymore, but I don't want to touch the existing code much, so this is now wrapped in a very specific request parameter, which all old code now passes.
* Some of our datapoints have been modified to use the correct PHP API client methods that rely on the recommended API endpoints. We've been using several deprecated endpoints so far (e.g. https://developers.google.com/adsense/management/v1.4/reference/accounts/adclients/list must be used over https://developers.google.com/adsense/management/v1.4/reference/adclients/list).
* This PR includes some logic to fix REST API error responses from datapoints, which is also included in #1264 (this one's still awaiting review).

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by felixarntz** [comment]
@tofumatt This is ready for review. One thing I'll need to change is to use the `stringifyObject` function in the `getReport` functionality, instead of `JSON.stringify` which I temporarily relied on for now. This change can be made once #1334 is merged.

File: assets/js/modules/adsense/datastore/__fixtures__/accounts.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[
+	{
+    "creationTime":"1382025063000",
+    "id":"pub-2833782679114991",
+    "kind":"adsense#account",
+    "name":"Test Account",
+    "premium":false,
+    "timezone":"Europe\/Berlin"
```
Comment (tofumatt) [inline assets/js/modules/adsense/datastore/__fixtures__/accounts.json:8]: Is that `\` intentional here? Seems odd? 🤷‍♂

File: assets/js/modules/adsense/datastore/alerts.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	 * Gets all Google AdSense alerts for this account.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state     Data store's state.
+	 * @param {string} accountID The AdSense Account ID to fetch alerts for.
+	 * @return {?Array.<Object>} An array of AdSense alerts; `undefined` if not loaded.
+	 */
+	getAlerts( state, accountID ) {
+		invariant( accountID, 'accountID is required.' );
```
Comment (tofumatt) [inline assets/js/modules/adsense/datastore/alerts.js:194]: As we've done elsewhere, we likely want to allow `undefined` here for selectors than may not have an `accountID` available yet. `null` shouldn't be allowed, because we should either pass a value here or just return `undefined` when `undefined` is supplied.

This is needed because our hooks need to run in a repeatable order inside components, so instead of doing:

```js
const accountID = useSelect(getAccountID);
if ( accountID ) {
  useSelect( select => select.getAlerts( accountID ) );
}
```

we have to allow for:

```js
const accountID = useSelect(getAccountID); // might return `undefined`
useSelect( select => select.getAlerts( accountID ) );
```

even if `accountID` is undefined.

File: assets/js/modules/adsense/setup/adsense-setup.js [hunk 3]
```diff
@@ -61,7 +61,7 @@ class AdSenseSetupWidget extends Component {
 
 	async getAccounts() {
 		try {
-			const responseData = await data.get( TYPE_MODULES, 'adsense', 'accounts' );
+			const responseData = await data.get( TYPE_MODULES, 'adsense', 'accounts', { maybeSetAccount: true } );
```
Comment (tofumatt) [inline assets/js/modules/adsense/setup/adsense-setup.js:64]: What's the reason for this change here? I didn't see it mentioned in the PR and it's a bit of a confusing param name 😅

File: assets/js/modules/adsense/util/parsing.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import { isValidAccountID, isValidClientID } from './validation';
+
+/**
```
Comment (tofumatt) [inline assets/js/modules/adsense/util/parsing.js:24]: It would be good to add some unit tests for these functions; they look straightforward but it'd be good to test them and make sure we're covering any edge cases, unexpected inputs, etc. 😄

File: assets/js/modules/adsense/util/validation.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Checks if the given account ID appears to be a valid AdSense account.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|number} accountID Account ID to test.
+ * @return {boolean} True if the given account ID is valid, false otherwise.
+ */
+export function isValidAccountID( accountID ) {
+	return typeof accountID === 'string' && accountID.match( /^pub-\d+$/ );
```
Comment (tofumatt) [inline assets/js/modules/adsense/util/validation.js:28]: This could use `parseAccountID` and check for a match instead of duplicating the regex. 🤷‍♀

File: assets/js/modules/adsense/datastore/alerts.js (hunk 2 above)
Comment (felixarntz) [reply assets/js/modules/adsense/datastore/alerts.js:194]: As discussed, let's add checks to all selectors that have required parameters:
* If any of the required parameters is `undefined`, immediately return `undefined`.
* That should be added to every such selector _and_ resolver.

cc @aaemnnosttv @tofumatt

File: assets/js/modules/adsense/setup/adsense-setup.js (hunk 3 above)
Comment (felixarntz) [reply assets/js/modules/adsense/setup/adsense-setup.js:64]: @tofumatt I mentioned in the "Relevant technical choices" section. The API route in PHP was running some ugly side effect logic, which I've now wrapped into this confusing name parameter (confusing intentionally because it shouldn't be used). I've added it to all existing code calling that route so that the behavior doesn't change, while preventing that side effect to run for our refactored code.

File: assets/js/modules/adsense/util/validation.js (hunk 5 above)
Comment (felixarntz) [reply assets/js/modules/adsense/util/validation.js:28]: It's a different regex, `parseAccountID` is for parsing the account ID from a client ID.
